x/ecocredit/server/marketplace/buy_direct.go

```diff
@@ -0,0 +1,81 @@
+package marketplace
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit/marketplace"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server/utils"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// BuyDirect allows for the purchase of credits directly from sell orders.
+func (k Keeper) BuyDirect(ctx context.Context, req *marketplace.MsgBuyDirect) (*marketplace.MsgBuyDirectResponse, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	buyerAcc, err := sdk.AccAddressFromBech32(req.Buyer)
+	if err != nil {
+		return nil, err
+	}
+	sellOrder, err := k.stateStore.SellOrderTable().Get(ctx, req.SellOrderId)
+	if err != nil {
+		return nil, fmt.Errorf("sell order %d: %w", req.SellOrderId, err)
+	}
+	if req.DisableAutoRetire && !sellOrder.DisableAutoRetire {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("cannot disable auto retire when purchasing credits " +
+			"from a sell order that does not have auto retire disabled")
+	}
+	batch, err := k.coreStore.BatchInfoTable().Get(ctx, sellOrder.BatchId)
+	if err != nil {
+		return nil, sdkerrors.ErrIO.Wrapf("error getting batch id %d: %s", sellOrder.BatchId, err.Error())
+	}
+	ct, err := utils.GetCreditTypeFromBatchDenom(ctx, k.coreStore, k.paramsKeeper, batch.BatchDenom)
+	if err != nil {
+		return nil, err
+	}
+	creditOrderQty, err := math.NewPositiveFixedDecFromString(req.Quantity, ct.Precision)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that bid price and ask price denoms match
+	market, err := k.stateStore.MarketTable().Get(ctx, sellOrder.MarketId)
+	if err != nil {
+		return nil, fmt.Errorf("market id %d: %w", sellOrder.MarketId, err)
+	}
+	if req.PricePerCredit.Denom != market.BankDenom {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("price per credit denom does not match ask price denom: "+
+			" %s, expected %s", req.PricePerCredit.Denom, market.BankDenom)
+	}
+	// check that bid price >= sell price
+	sellOrderPricePerCredit, ok := sdk.NewIntFromString(sellOrder.AskPrice)
+	if !ok {
+		return nil, fmt.Errorf("could not convert %s to %T", sellOrder.AskPrice, sdk.Int{})
```

Comment:
i don't think it is, theres nothing from the request in these lines. sellOrder is pulled from state. i think if anything it would be some sort of internal/server error

Reply:
OK